Python/Product/Analysis/Interpreter/Ast/AstPythonInterpreterFactory.cs

```diff
@@ -441,30 +446,53 @@ bool useDefaultDatabase
 
         #region Module Imports
 
-        public enum TryImportModuleResult {
+        /// <summary>
+        /// Determines whether the specified directory is an importable package.
+        /// </summary>
+        public bool IsPackage(string directory) {
+            return ModulePath.PythonVersionRequiresInitPyFiles(Configuration.Version) ?
+                !string.IsNullOrEmpty(ModulePath.GetPackageInitPy(directory)) :
+                Directory.Exists(directory);
+        }
+
+        public enum TryImportModuleResultCode {
             Success,
             ModuleNotFound,
             NeedRetry,
             NotSupported,
             Timeout
         }
 
+        public struct TryImportModuleResult {
+            public TryImportModuleResultCode Status;
+            public IPythonModule Module;
+
+            public TryImportModuleResult(IPythonModule module) {
+                Status = module == null ? TryImportModuleResultCode.ModuleNotFound : TryImportModuleResultCode.Success;
+                Module = module;
+            }
+
+            public static TryImportModuleResult ModuleNotFound => new TryImportModuleResult { Status = TryImportModuleResultCode.ModuleNotFound };
+            public static TryImportModuleResult NeedRetry => new TryImportModuleResult { Status = TryImportModuleResultCode.NeedRetry };
+            public static TryImportModuleResult NotSupported => new TryImportModuleResult { Status = TryImportModuleResultCode.NotSupported };
+            public static TryImportModuleResult Timeout => new TryImportModuleResult { Status = TryImportModuleResultCode.Timeout };
+        }
+
         public sealed class TryImportModuleContext {
             public IPythonInterpreter Interpreter { get; set; }
             public ConcurrentDictionary<string, IPythonModule> ModuleCache { get; set; }
-            public int Timeout { get; set; } = 5000;
             public IPythonModule BuiltinModule { get; set; }
-            public Func<string, Task<ModulePath?>> FindModuleInUserSearchPathAsync { get; set; }
+            public Func<string, CancellationToken, Task<ModulePath?>> FindModuleInUserSearchPathAsync { get; set; }
```

Comment:
I don't understand the idea of this delegate. Referenced method itself is called only from `AstPythonInterpreterFactory`, why it is in the `AstPythonInterpreter`?

Reply:
It uses private properties and private state of `AstPythonInterpreter`, and may be implemented differently for different interpreters.

Reply:
It is very hard to read. Ok, it is out of this PR scope.

Reply:
The only alternative is defining another public interface, and since it's meant to be private to this class, that doesn't really make sense. This gets worse if other callers (i.e. us) decide to start partially (and incorrectly) resolving methods through a public interface that is meant for one precise purpose.